sql/src/main/java/io/crate/executor/transport/SnapshotRestoreDDLDispatcher.java

```diff
@@ -169,77 +173,96 @@ public void onFailure(Exception e) {
+    static CompletableFuture<ResolveIndicesAndTemplatesContext> resolveIndexNames(List<RestoreSnapshotAnalyzedStatement.RestoreTableInfo> restoreTables, boolean ignoreUnavailable,
+                                                                                  TransportGetSnapshotsAction getSnapshotsAction, String repositoryName) {
+        ResolveIndicesAndTemplatesContext resolveIndicesAndTemplatesContext = new ResolveIndicesAndTemplatesContext();
         List<RestoreSnapshotAnalyzedStatement.RestoreTableInfo> toResolveFromSnapshot = new ArrayList<>();
         for (RestoreSnapshotAnalyzedStatement.RestoreTableInfo table : restoreTables) {
             if (table.hasPartitionInfo()) {
-                resolvedIndices.add(table.partitionName().asIndexName());
+                resolveIndicesAndTemplatesContext.addIndex(table.partitionName().asIndexName());
+                resolveIndicesAndTemplatesContext.addTemplate(PartitionName.templateName(table.tableIdent().schema(), table.tableIdent().name()));
             } else if (ignoreUnavailable) {
                 // If ignoreUnavailable is true, it's cheaper to simply return indexName and the partitioned wildcard instead
                 // checking if it's a partitioned table or not
-                resolvedIndices.add(table.tableIdent().indexName());
-                resolvedIndices.add(PartitionName.templateName(table.tableIdent().schema(), table.tableIdent().name()) + "*");
+                resolveIndicesAndTemplatesContext.addIndex(table.tableIdent().indexName());
+                // For the case its a partitioned table we restore all partitions and the templates
+                String templateName = PartitionName.templateName(table.tableIdent().schema(), table.tableIdent().name());
+                resolveIndicesAndTemplatesContext.addIndex(templateName + "*");
+                resolveIndicesAndTemplatesContext.addTemplate(templateName);
             } else {
                 // index name needs to be resolved from snapshot
                 toResolveFromSnapshot.add(table);
             }
         }
         if (toResolveFromSnapshot.isEmpty()) {
-            return CompletableFuture.completedFuture(ImmutableList.copyOf(resolvedIndices));
+            return CompletableFuture.completedFuture(resolveIndicesAndTemplatesContext);
         }
-        final CompletableFuture<List<String>> f = new CompletableFuture<>();
+        final CompletableFuture<ResolveIndicesAndTemplatesContext> f = new CompletableFuture<>();
         getSnapshotsAction.execute(
             new GetSnapshotsRequest(repositoryName),
-            new ResolveFromSnapshotActionListener(f, toResolveFromSnapshot, resolvedIndices)
+            new ResolveFromSnapshotActionListener(f, toResolveFromSnapshot, resolveIndicesAndTemplatesContext)
         );
         return f;
     }
 
+    static class ResolveIndicesAndTemplatesContext {
+
+        Collection<String> resolvedIndices = new HashSet<>();
```

Comment:
as i use it in the tests i need it package private. Made it final and added the @VisibleForTesting annotation to the class.